pkg/agent/openflow/pipeline_windows.go

```diff
@@ -311,3 +345,51 @@ func (c *client) l3FwdFlowToRemoteViaRouting(localGatewayMAC net.HardwareAddr, r
 	}
 	return []binding.Flow{c.l3FwdFlowToRemoteViaGW(localGatewayMAC, *peerPodCIDR, category)}
 }
+
+// clusterIPHostNetworkFlows installs flows needed for a ClusterIP Service which is accessed from Windows host.
+func (c *client) clusterIPHostNetworkFlows(svcIP net.IP, gwIP net.IP, nodeIP net.IP, category cookie.Category) (flows []binding.Flow) {
+	flows = append(flows,
+		// Hostnetwork case needs SNAT so the SNAT mark is set.
+		c.pipeline[l3ForwardingTable].BuildFlow(priorityLow).MatchProtocol(binding.ProtocolIP).
+			MatchSrcIP(gwIP).MatchCTMark(ServiceCTMark, nil).
+			MatchRegRange(int(marksReg), markTrafficFromGateway, trafficFromRegRange).
+			Action().LoadRegRange(int(marksReg), snatDefaultMark, snatMarkRange).
+			Action().ResubmitToTable(l2ForwardingCalcTable).
+			Cookie(c.cookieAllocator.Request(category).Raw()).
+			Done(),
+		// SNAT: antrea-gw0 -> NodeIP.
+		c.pipeline[conntrackCommitTable].BuildFlow(priorityHigh).MatchProtocol(binding.ProtocolIP).
+			MatchSrcIP(gwIP).MatchCTMark(ServiceCTMark, nil).
+			MatchCTStateNew(true).MatchCTStateTrk(true).
+			MatchRegRange(int(marksReg), snatDefaultMark, snatMarkRange).
+			MatchInPort(config.HostGatewayOFPort).
+			Action().CT(true, L2ForwardingOutTable, ctZoneSNAT).SNAT(
+			&binding.IPRange{StartIP: nodeIP, EndIP: nodeIP},
+			nil).
+			LoadToMark(snatCTMark).CTDone().
+			Cookie(c.cookieAllocator.Request(category).Raw()).
+			Done(),
+		c.pipeline[conntrackCommitTable].BuildFlow(priorityHigh).MatchProtocol(binding.ProtocolIP).
+			MatchSrcIP(gwIP).MatchCTMark(ServiceCTMark, nil).
+			MatchCTStateNew(false).MatchCTStateTrk(true).
+			MatchRegRange(int(marksReg), snatDefaultMark, snatMarkRange).
+			Action().CT(false, L2ForwardingOutTable, ctZoneSNAT).NAT().CTDone().
+			Cookie(c.cookieAllocator.Request(category).Raw()).
+			Done(),
+		// The request packet is received from antrea-gw0 and will be sent back to antrea-gw0 for routing, so
+		// "in_port" is used as output which is required by OVS.
+		c.pipeline[L2ForwardingOutTable].BuildFlow(priorityHigh).MatchProtocol(binding.ProtocolIP).
+			MatchRegRange(int(marksReg), trafficFromGatewayAndSNATMark, trafficFromAndSNATMarkRange).
+			Action().OutputInPort().
+			Cookie(c.cookieAllocator.Request(category).Raw()).
+			Done(),
+		// For IP binding no-hostnetwork Service, the request IP should be SNATed to antrea-gw0 IP.
```

Comment:
Could you explain why it must be SNAT'd to gw0 IP? And you code actuall SNAT to svcIP not gw0 IP?

Probably rephrase "For IP binding no-hostnetwork Service". I did not understand the sentence.

Reply:
Sorry, this is a mistake and it should be gw0 IP. The case is actually the one you mentioned "the host uses another IP to access the Service". Since this local another IP is not available to remote pod/node, it is better SNAT'd to gw0 IP.